include/openthread/tcp.h

```diff
@@ -0,0 +1,589 @@
+ * @retval OT_ERROR_FAILED  Failed to shut down the desired side(s) of the connection.
+ */
+otError otTcpActiveShutdown(otInstance *aInstance, otTcpActiveSocket *aSocket, otTcpActiveShutdownHow aHow);
+
+/**
+ * Forcibly ends the TCP connection associated with this active socket. This
+ * immediately makes the active socket free for use for another connection and
+ * empties the send and receive buffer, transferring ownership of any data
+ * provided by the application in @p otTcpActiveSendByReference calls back to
+ * the application. The socket's callbacks and memory for the receive buffer
+ * remain associated with the socket.
+ *
+ * @param[in]  aInstance  A pointer to the OpenThread instance to which the socket belongs.
+ * @param[in]  aSocket    A pointer to the active TCP socket structure representing the TCP endpoint to abort.
+ *
+ * @retval OT_ERROR_NONE    Successfully closed the active TCP socket.
+ * @retval OT_ERROR_FAILED  Failed to close the active TCP socket.
+ */
+otError otTcpActiveAbort(otInstance *aInstance, otTcpActiveSocket *aSocket);
+
+/**
+ * Closes this active TCP socket. If it corresponds to a live TCP connection,
+ * the connection is terminated unceremoniously (as in @p otTcpActiveAbort).
+ * All resources the application has provided for this socket (linked buffers
+ * for the send buffer, memory for the receive buffer, the @p otTcpActiveSocket
+ * structure itself, etc.) are immediately returned to the application. This is
+ * different from the classic BSD API, in which the TCP connection may outlive
+ * the socket.
+ *
+ * @param[in]  aInstance  A pointer to the OpenThread instance to which the socket belongs.
+ * @param[in]  aSocket    A pointer to the active TCP socket structure to close.
+ *
+ * @retval OT_ERROR_NONE    Successfully closed the active TCP socket.
+ * @retval OT_ERROR_FAILED  Failed to close the active TCP socket.
+ */
+otError otTcpActiveClose(otInstance *aInstance, otTcpActiveSocket *aSocket);
+
+struct otTcpPassiveSocket;
+typedef struct otTcpPassiveSocket otTcpPassiveSocket;
+
+/**
+ * This callback indicates that an incoming connection that matches this
+ * passive socket has arrived. The application should provide (via the
+ * callback's return value) a pointer to the @p otTcpActiveSocket into which
+ * to accept the incoming connection. If the application returns OT_TCP_DEFER,
+ * then OpenThread silently ignores the connection establishment request; the
+ * callback will be called again the next time that the connection
+ * establishment request is submitted. This is valuable if resources are not
+ * available to accept the connection, but they may be available at the time of
+ * a future attempt. If the application returns OT_TCP_REFUSE, then OpenThread
+ * sends a "connection refused" message to the host that attempted to establish
+ * a connection.
+ */
+typedef otTcpActiveSocket *(*otTcpAcceptReady)(void *aContext, otTcpPassiveSocket *aSocket, const otSockAddr *aPeer);
```

Comment:
How do we pass `otTcpActiveOpenArgs` to this active socket?

Reply:
The active socket should be opened already (passing `otTcpActiveOpenArgs` at the time it is opened) before a pointer to it is provided to the TCP implementation via the `otTcpAcceptReady` callback.

Reply:
I see. Should we make it clear in `otTcpAcceptReady` doxygen?

Reply:
Yes, I agree it's not clear from the documentation. I'll push an update.

Reply:
I pushed an update clarifying this.